Introduce `caml_update_dummy_lazy` to avoid value-letrec bugs due to Forward shortcuts

gasche opened the pull request:
This is a proposed fix for #13930, where @lthls observes a segfault on the following program:

```ocaml
let f x =
  let rec l =
    let v = lazy x in
    Gc.minor ();
    v
  in
  l
let lazy_one = f 1
```

Recursive values use a tying-the-knot approach where they are first pre-allocated as "dummy" values using `caml_alloc_dummy`, and then updated to their final value using `caml_update_dummy`, which assumes that the final value has the same size as the dummy and copies its tags and fields into the dummy. (We only accept recursive values such that the size can be computed statically before preallocation.)

The problem on this example is that the type-checker predicts that the final value will be a `Forward` block (of size 1), and it indeed starts this way but then a GC shortcuts it to the underlying value, which can arbitrary (of any size, or an immediate).

The fix proposes a dedicated `caml_update_dummy_lazy` primitive for lazy values. When the final value does not have a lazy tag, we know that it comes from a Forward tag that was shortcut, and we can perform the update correctly.

In the process of writing this fix, I:
- migrated `caml_{alloc,update}_dummy` from alloc.c to obj.c
- removed now-dead code in `caml_update_dummy` that was dealing with the Infix_tag case, before @lthls rewrote value-letrec compilation to never use dummies on functions.

gasche committed:
runtime: migrate caml_{alloc,update}_dummy from alloc.c to obj.c

gasche committed:
caml_update_dummy: the [Infix_tag] case is dead code

We stopped using dummies for values at a function type in OCaml 5.2,
and the matching [caml_alloc_dummy_infix] function was removed in
cd48dc0ac20f58fa460b55fafb75784d5a9d62f3.

gasche committed:
[minor] value_rec_compiler: refactor alloc/update calls

gasche committed:
recursive values: introduce [caml_update_dummy_lazy]

lthls committed:
regression test for lazy shortcutting in recursive values

gasche commented:
Note: if we go this route, we need to let people who have alternative runtimes (notably js_of_ocaml) know of the new `caml_update_dummy_lazy` function so that they can implement it as well.

(cc @hhugo : I'm not completely sure that this is the approach we will take in the end, but this is advance notice)

lthls commented:
I will do a proper review next week, but I was wondering about two things that might still be problematic even after this PR. Consider the following code:
```ocaml
let rec v =
  let l = lazy (expensive computation) in
  let () = maybe_force_in_another_domain l in
  l
```
(Note that `v` must not be actually recursive for this case to be accepted).

The first thing I'm worried about is if there is a race between the reads from `l` to update the dummy, and the writes from the other domain forcing `l`. I think that we could, in theory, end up reading a `Lazy` tag, but by the time we copy the fields the value has been forwarded, so we store the forwarded value in the block instead of the closure.
Additionally, if the lazy block had tag `Forcing` at the time where `caml_update_dummy_lazy` is called, we could end up creating a block with `Forcing` tag that nobody is actually forcing, which sounds a bit problematic.
There is also the potential problem of duplicating a non-forced lazy block, which means it could get forced twice independently, but I consider this in the same category as duplicating mutable blocks (annoying, but not likely to cause segfaults).

What do you think ? Do you want to try to handle the issue with races in this PR ? The rest seems out of the scope of this PR to me.

gasche commented:
If I remember correctly, shortcutting of Forward blocks happens during the minor GC. In particular, if we observe a non-shortcut lazy block (with an atomic read that synchronizes enough, as `obj_tag` does), then until the next STW poll point no shortcutting can happen and we will keep observing lazy tags. This reasoning is documented in CamlinternalLazy.force_gen.

The tags can be change concurrently in the case of a forcing race, though. Lazy can become Forcing, and Forcing can become either Lazy (on error) or Forward (on success). The payload is the same for Lazy and Forcing block (a function), and has a different type for Forward (its result).

I think that the following strategy would be okay:
- If we see Lazy, copy the the argument and *then* read the tag again (with synchronization) to check that it is still Lazy. If we have seen it Lazy (with synchronization) before and after the field read, then we know that the field has the expected type -- it is a function.
- If we see Forward, we know the tag will not change again before the next GC, so we can copy as currently.

In the details, we can be permissive or fail-loudly in presence of concurrent updates:
- we can treat Forcing just like Lazy, or fail with Lazy.Undefined because we can tell we are observing (forbidden) concurrent modification
- when we check the thunk tag a second time after copying the field, if we see Forward, we can fail with Lazy.Undefined (we observed a concurrent modification), or we can forget about ever seeing Lazy/Forcing and silently move to the Forward strategy

I would be inclined to implement the fail-loudly behavior which is more in line with the rest of the runtime.

In any of these choices there is always a risk that the thunk function is called several times because we duplicated a non-forced lazy block, as you point out. I agree that this is not too bad -- it's better when we can fail instead of silently doing something strange, but I don't think we can do much better with the current representation of lazy.

gasche committed:
lazy + letrec: insert indirections on lazy update to avoid concurrency issues

gasche committed:
Changes

gasche commented:
@lthls I thought of a solution for the problem of concurrent forcing that is more correct than what I discussed above (it never introduces forcing-related failures), and integrated it in the present PR. It pessimizes slightly value-letrecs involving lazy thunks, in a way that I think is acceptable.

My suggestion above was along the lines of handling the return of the value-letrec as a forcing event: we inspect the thunk, and if we notice that someone else is forcing it concurrently then we fail. After thinking more about this I decided that it is problematic, because we do tell our users that *forcing races* are programming errors, but here there is no forcing races, one domain is forcing the value and the other is just trying to copy the thunk. In your example, in theory, a user could be forcing the thunk in a different domain, and then doing some synchronization before forcing the recursively-defined thunk in the current domain, and they would not consider this usage racy.

The solution I propose is, instead of writing `newval` in the dummy, to write a *fresh* proxy thunk defined as `lazy (Lazy.force newval)`. We can do this without touching `newval` at all, and in particular this is correct even if another domain is forcing `newval`. I implemented this via a new `CamlinternalLazy.indirect` function (it would be difficult to create this thunk in C, as creating closures is difficult there). In fact, the function only creates this proxy thunk if `newval` is not evaluated yet; if it is already a forced value, we return it directly.

In other words, we are moving back to a tying-the-knot phase that inserts an indirection, instead of the `caml_update_dummy` scheme that tried to update the final value in-place. I think that this is acceptable in this case. The reason why we want to avoid indirections in general is that they pollute the representation of the recursive value after evaluation, all future clients pay the cost of traversing the extra indirections. Lazy thunks have the good property that the indirection will only be paid once, when forcing the recursively-defined chunk, and gets shortcut after. So I believe this is acceptable.

lthls commented:
I have an alternate proposal (that I will submit as soon as the testsuite passes) that would fix the same issues using a different approach. The idea is to move the `caml_update_dummy` calls earlier, in a way that makes it impossible to have duplicated blocks.

On an example:
```ocaml
(* Before compilation *)
let rec v =
  let r = ref 0 in
  escape r;
  r

(* After compilation *)
let v = alloc_dummy 1 in
let r = (caml_update_dummy v (ref 0); v) in
escape r;
ignore r
```

I think it's a good idea in general, but it's a more complex patch. What do you think ? Should we first try to get this PR in, even though most of it will be redundant if we merge my proposal ?

gasche commented:
@lthls commented in #13938:

> the PR seems to miss the actual addition of [the camlinternalLazy.indirect] primitive

I see it in the diff, as a new function in stdlib/camlinternalLazy.ml. Or maybe I misunderstood your remark?

lthls commented:
> @lthls commented in #13938:
> 
> > the PR seems to miss the actual addition of [the camlinternalLazy.indirect] primitive
> 
> I see it in the diff, as a new function in stdlib/camlinternalLazy.ml. Or maybe I misunderstood your remark?

I looked for it and did not see it, but now I do -- maybe I forgot to refresh the page. Sorry about the noise.

gasche commented:
Note: you mentioned that you are less convinced than I am by the indirection. I'm happy to remove this part of the PR for now and send it separately, if it can help.

lthls commented on lambda/value_rec_compiler.ml:
```diff
@@ -49,6 +49,7 @@ open Lambda
 type block_size =
   | Regular_block of int
   | Float_record of int
+  | Lazy_block of int
```
I wouldn't have put the size here, since it can only be 1 for lazy blocks, but if you prefer it your way feel free to keep it

gasche replied:
I did wonder about this, but we have discussed changes to the Lazy implementation that would change its payload size. (The problem today is that we overwrite the code pointer to write the final result, which makes for tricky concurrent reasoning, the runtime would be noticeably simpler if we reserved two words, one for the thunk and one for the result.) I suspect that it would be slightly easier to adapt the code if we keep this an explicit parameter here. 

Another option would be to move to a `caml_alloc_dummy_lazy` as I considered in the past; currently only the `update` part is type-specialized, because I can use the standard dummy allocator. In an earlier version I tried to put a more precise tag on dummy values, but then I realized that preallocating a dummy block with `Forward_tag` was not such a great idea.

lthls commented on lambda/value_rec_compiler.ml:
```diff
@@ -699,6 +700,68 @@ let update_prim =
+             This removes indirections on terms defined as lazy thunks
+             at the toplevel: [let rec x = lazy ...] *)
+          newval
+        | _ -> compile_indirect newval
```
Note that #13938 should make this case unreachable (if we end up merging it)

gasche replied:
I find it not-so-easy to convince myself that this is indeed the case -- I think that it works because you reason about the current implementation of the size-computation in Value_rec_compiler, we can check all cases that return a Lazy_block values and see that they are all fresh creation, I suppose. If this was obvious I would suggest (you have been too polite to do it yourself) to remove the `indirect` stuff entirely, but I think that having it for now and removing it when we merge #13938 is probably the safer choice.

lthls commented on runtime/obj.c:
```diff
@@ -300,3 +300,81 @@ struct queue_chunk {
+    size = Wosize_val(newval);
+    CAMLassert (size == Wosize_val(dummy));
+    /* See comment above why this is safe even if [tag == Closure_tag]
+       and some of the "values" being copied are actually code pointers. */
```
This comment should be either updated or removed (it refers to a "comment above" that has just been deleted). Given that we cannot have code pointers anymore here, I would go with removing it.

lthls commented on runtime/obj.c:
```diff
@@ -300,3 +300,81 @@ struct queue_chunk {
+  default:
+    CAMLassert (Wosize_val(dummy) == 1);
+    caml_modify(&Field(dummy, 0), newval);
+    Unsafe_store_tag_val(dummy, Forward_tag);
```
It probably doesn't matter, but `caml_update_dummy` writes the tag before the contents, so I would have done the same here for consistency.

gasche replied:
I hesitated, but I don't want to have to worry about an intermediate state where the runtime sees `Forward_tag` with a type-invalid payload.

lthls commented on testsuite/tests/letrec-compilation/lazy_.ml:
```diff
@@ -1,5 +1,19 @@
 (* TEST *)
 
-let rec c = lazy (0 + d) and d = 3;;
+let rec c = lazy (0 + d) and d = 42;;
```
Nitpick , but I don't think it is a good idea to use the same constant twice in a test (a bug might cause you to use the wrong value, but since you have used the same constant you don't detect it).

lthls reviewed (approved):
I have left some comments, but I'm approving the general idea and implementation.

gasche commented:
Thanks! I will take your comments into account and update the PR.

lthls commented:
I almost forgot: to fix #13931, I need to be able to find the classification of previously compiled bindings by looking at the `caml_alloc_dummy` primitive, so I would like you to introduce `caml_alloc_dummy_lazy` (which just calls `caml_alloc_dummy`) for that. With the current approach, we risk using `caml_update_dummy` instead of `caml_update_dummy_lazy` for some lazy bindings.

gasche committed:
use a dedicated [caml_alloc_dummy] primitive

Suggested-by: Vincent Laviron <vincent.laviron@gmail.com>

gasche commented:
I rebased the PR to take your comments into account, and in particular I added a new commit that adds a dedicated `caml_alloc_dummy_lazy` primitive as you suggest. (I decided to simultaneously remove the `size` parameter from `Lazy_block`, as I think the two changes are related.)

I'm not fond of the idea that the middle-end relies on detecting specific C function names and give them a semantic. I think it may be time to create a proper `Pmakedummyblock` primitive instead. I'm interested in your opinion but I think this could be done as a separate change. (For example, you could do it yourself if/when you send a PR dedicated to fixing #13931.)

gasche commented:
(My plan is to give @lthls the opportunity for a new round of feedback, in particular on the new commit, before merging.)

lthls commented:
> I'm not fond of the idea that the middle-end relies on detecting specific C function names and give them a semantic. I think it may be time to create a proper `Pmakedummyblock` primitive instead. I'm interested in your opinion but I think this could be done as a separate change.

I completely agree with you on this. It would also allow us to recover a few very specific optimisations in flambda that were lost when we switched to `Value_rec_compiler`, if we wanted to. I think it's out of scope for this PR, and I suspect it's too big a change to be included in a quick fix for #13931, but it would make sense as part of a bigger change like #13938.

lthls commented:
I have looked at the new version, and I don't have anything to add.

gasche commented:
Merging, then! Thanks for the excellent review as always.
